### ISSUE TITLE
Inconsistent as_tensor converts device when given a dtype

### ISSUE THREAD

**Issue by florensacc** [pytorch/pytorch#45367]
## 🐛 Bug

When calling `torch.as_tensor` on a tensor `data` that is on the GPU it stays there, but if we specify `dtype=data.dtype` it's copied to the CPU!!

Might be related to https://github.com/pytorch/pytorch/issues/14996

## To Reproduce

```
data = torch.ones(3, device=torch.device("cuda"))

data1 = torch.as_tensor(data)
print(data1.device)

data2 = torch.as_tensor(data, dtype=data.dtype)
print(data2.device)
```

prints:
```
cuda:0
cpu
```

## Expected behavior

Both should be `cuda:0`.

## Environment

PyTorch version: 1.5.0
Is debug build: False
CUDA used to build PyTorch: 10.2
ROCM used to build PyTorch: N/A

OS: Ubuntu 18.04.3 LTS (x86_64)
GCC version: (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0
Clang version: Could not collect
CMake version: version 3.10.2

Python version: 3.7 (64-bit runtime)
Is CUDA available: True
CUDA runtime version: 10.0.130
GPU models and configuration:
GPU 0: GeForce GTX 1080 Ti
GPU 1: GeForce GTX 1080 Ti
GPU 2: GeForce GTX 1080 Ti
GPU 3: GeForce GTX 1080 Ti

Nvidia driver version: 440.64.00
cuDNN version: /usr/lib/x86_64-linux-gnu/libcudnn.so.7.6.5
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] msgpack-numpy==0.4.4.3
[pip3] numpy==1.16.1
[pip3] numpy-quaternion==2020.8.13.21.53.18
[pip3] numpy-stl==2.10.1
[pip3] torch==1.5.0
[pip3] torchvision==0.6.0
[conda] msgpack-numpy             0.4.4.3                  pypi_0    pypi
[conda] numpy                     1.16.1                   pypi_0    pypi
[conda] numpy-quaternion          2020.8.13.21.53.18          pypi_0    pypi
[conda] numpy-stl                 2.10.1                   pypi_0    pypi
[conda] torch                     1.5.0                    pypi_0    pypi
[conda] torchvision               0.6.0                    pypi_0    pypi

cc @ngimel

**Comment by gchanan** [comment]
looks like this got fixed on master:

```
>>> import torch
>>> torch.__version__
'1.7.0a0+8232166'
>>> data = torch.ones(3, device=torch.device("cuda"))
>>> data1 = torch.as_tensor(data)
>>> print(data1.device)
cuda:0
>>> data2 = torch.as_tensor(data, dtype=data.dtype)
>>> print(data2.device)
cuda:0
```

I confirmed this is an issue on 1.6.0.

**Comment by ngimel** [comment]
Fixed in #41984 and the fix is bc-breaking.

**Comment by mtikekar** [comment]
Hi @ngimel, I believe this is a wrong fix (or at least, inconsistent with the documentation) According to: https://pytorch.org/docs/master/generated/torch.as_tensor.html#torch.as_tensor

> torch.as_tensor(data, dtype=None, device=None) → Tensor
>   ...
>   device (torch.device, optional) – the desired device of returned tensor. Default: if None, uses the current device for the default tensor type (see torch.set_default_tensor_type()). device will be the CPU for CPU tensor types and the current CUDA device for CUDA tensor types.

In both the cases above, `device` is `None`, which means it should use the cpu device as that is the default tensor type (unless someone explicitly runs something like `torch.set_default_tensor_type(torch.cuda.FloatTensor)` first.)

So, to be consistent with the documentation, the expected behavior in both cases is `cpu` and not `cuda:0`. Or, the documentation should be updated to something like:

> device (torch.device, optional) - the desired device of returned tensor. Default: if None __and input is a numpy array__, uses the current device for the default tensor type (see torch.set_default_tensor_type()). device will be the CPU for CPU tensor types and the current CUDA device for CUDA tensor types. If None __and input is a tensor__, uses the device of the input tensor
